### PR TITLE
small cleanup in monoids

### PR DESCRIPTION
**Pull request by fchapoton**
just making many small changes in monoids

some scripted using ruff, about string formatting

some hand-made, about pep and code details

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.

### CODE REVIEW

File: src/sage/monoids/free_monoid_element.py [hunk 1]
```diff
@@ -159,9 +159,9 @@ def _latex_(self):
             g = x[int(v[i][0])]
             e = v[i][1]
             if e == 1:
-                s += "%s " % (g,)
+                s += f"{g} "
             else:
-                s += "%s^{%s}" % (g, e)
+                s += f"{g}^{{{e}}}"
```
Comment (mkoeppe) [inline src/sage/monoids/free_monoid_element.py:164]: f-strings for latex get weird pretty quickly. maybe we need a general helper function in `sage.typeset` for putting things into braces?

File: src/sage/monoids/free_monoid_element.py [hunk 2]
```diff
@@ -373,9 +372,9 @@ def to_word(self, alph=None):
         gens = self.parent().gens()
         if alph is None:
             alph = gens
-        alph = [str(_) for _ in alph]
-        W = Words(alph)
-        return W(sum([ [alph[gens.index(i[0])]] * i[1] for i in list(self) ], []))
+        alph = [str(c) for c in alph]
+        W = Words(alph, infinite=False)
+        return W(sum([[alph[gens.index(i[0])]] * i[1] for i in iter(self)], []))
```
Comment (mkoeppe) [inline src/sage/monoids/free_monoid_element.py]: not just `for i in self`?

File: src/sage/monoids/string_ops.py [hunk 3]
```diff
@@ -94,24 +98,21 @@ def coincidence_discriminant(S,n=2):
...
-    if not isinstance(S,(list,tuple)):
+    if not isinstance(S, (list, tuple)):
         raise TypeError("Argument S (= %s) must be a list or tuple" % S)
     if n != 2:
         raise ValueError("Argument n (= %s) is only implemented for n = 2" % n)
-    truth = True
-    for bool in ( isinstance(c,(str,StringMonoidElement)) for c in S ):
-        truth = truth and bool
+    truth = all(isinstance(c, (str, StringMonoidElement)) for c in S)
     if not truth:
```
Comment (mkoeppe) [inline src/sage/monoids/string_ops.py]:
```suggestion
    if not all(isinstance(c, (str, StringMonoidElement)) for c in S):
```

File: src/sage/monoids/string_ops.py [hunk 4]
```diff
@@ -94,24 +98,21 @@ def coincidence_discriminant(S,n=2):
...
         raise TypeError("Argument S (= %s) must be a list of strings.")
-    for bool in ( len(c) == n for c in S ):
-        truth = truth and bool
+    truth = all(len(c) == n for c in S)
     if not truth:
```
Comment (mkoeppe) [inline src/sage/monoids/string_ops.py]:
```suggestion
    if not all(len(c) == n for c in S):
```

File: src/sage/monoids/trace_monoid.py [hunk 5]
```diff
@@ -818,8 +818,8 @@ def dependence_graph(self):
             sage: M.dependence_graph() == Graph({a:[a,b], b:[b], c:[c,b]})
             True
         """
-        return Graph(set(frozenset((e1, e2)) if e1 != e2 else (e1, e2)
-                         for e1, e2 in self.dependence()), loops=True,
+        return Graph({frozenset((e1, e2)) if e1 != e2 else (e1, e2)
+                         for e1, e2 in self.dependence()}, loops=True,
```
Comment (mkoeppe) [inline src/sage/monoids/trace_monoid.py]:
```suggestion
                      for e1, e2 in self.dependence()}, loops=True,
```

File: src/sage/monoids/string_ops.py (hunk 3 above)
Comment (grhkm21) [reply src/sage/monoids/string_ops.py]: Is it desirable to use `any(not *)` instead?

Comment (mkoeppe) [reply src/sage/monoids/string_ops.py]: both variants have the same short-circuiting semantics, and I'd say there's no difference in readability

**Review by mkoeppe — APPROVED** [review]
LGTM.

**Pull request merged**